### DISCUSSION TITLE
How to automate "Replace templates in the active file" ?

### DISCUSSION THREAD

**Discussion by doctorfree** [SilentVoid13/Templater#906] [Help] [answered]
Hi, I am new to Templater so maybe this is documented but I could not find it. I added a bunch of Templater templates to my vault documents using a shell script with a bunch of `sed` and `awk` and `grep`. It worked and made it easy to modify my vault documents in batch form. I have a lot of documents in this vault. I can open one of the documents modified in this way in Obsidian and use the Hotkey to "Replace templates in the active file". That works but it is going to take me quite a long time to do this manually for all the files.

Is there a way to automate this process? Something like "Replace templates in all files in vault" ? I looked at possibly using the Obsidian URI to construct a command line that would do this but I do not believe the URI command line interface can do anything more than open a document, not send it a Hotkey.

Maybe this is easy and I just do not know Templater well enough? Or maybe it is just not a supported workflow?

Any help greatly appreciated.

**Comment by AB1908** [comment, marked as answer]
You may want to redo some of this. Templater has a setting to resolve templates on new file creation but I suspect you have it turned off. Once you turn it on, try moving the files outside the vault and then move them back in. If that doesn't work, try recreating them via the script. If that still doesn't work, write a script to generate them in Templater itself and that should give you more than enough flexibility.

**Reply by doctorfree** [reply]
Thanks for the prompt reply. I had the setting to resolve templates on new file creation turned off. I turned it on following your suggestion, moved the files outside the vault and moved them back in. That sort of worked. All of the templates were resolved but some of them resolved to the value "undefined" even though the front matter included a valid definition for those. A bit perplexing.

I will perform some additional tests, perhaps moving a few at a time, to see how to complete this without error. At any rate, thanks for the help - moving documents out of the vault and back into the vault to fool Obsidian into thinking they are newly created documents seems to be a way to get Templater to do bulk template resolution.

**Comment by doctorfree** [comment]
I was able to successfully resolve all Templater templates by copying in the documents one at a time to a new location in the vault, sleeping a bit between copies. This is an ok one-time workaround for large bulk template ressolutions (for templates added in this manner, using text processing utilities outside of Obsidian/Templater).

However, it appears there may be some kind of race condition or other issue interfering with the processing of new notes in either Obsidian or Templater. Why is the sleep necessary? I am not yet familiar with the Obsidian plugin architecture so I do not have a good guess nor do I have a simple minimal test case that reproduces this problem. Ha! Not much help.

But, generally, I am satisfied with this procedure. Large bulk template resolution is hopefully a rare occurrence and the sleep only means I have to work on something else while that runs. Templater is a really great plugin. Thanks!

**Reply by AB1908** [reply]
Yeah there's some weird stuff happening and I still dunno why. Glad that you got it working though !

**Comment by lursyy** [comment]
It seems to me like there could be some race condition or other problem regarding parallel processing of the new files: in my case, some of the templates in file A got placed in file B and C, but not file A.

**Reply by lursyy** [reply]
For people looking for a workaround, here is what worked for me

```
#!/bin/zsh
find /path/to/source/root -type f | while read file; do 
  subdirpath="${file#/path/to/source/root/}"
  destfile="/path/to/target/root/$subdirpath"
  mkdir -p "$(dirname "$destfile")"
  cp "$file" "$destfile"
  sleep 1
done
```
